module/zfs/abd.c

```diff
@@ -0,0 +1,1542 @@
+	 *  allocate a scatter page vector.
+	 */
+	{ "scatter_page_vector_retry",		KSTAT_DATA_UINT64 },
+	/*
+	 *  The total number of retries encountered when attempting to
+	 *  allocate the pages to populate a scatter page vector.
+	 */
+	{ "scatter_page_alloc_retry",		KSTAT_DATA_UINT64 },
+	/*
+	 *  The total number of retries encountered when attempting to
+	 *  allocate the sg table for an ABD.
+	 */
+	{ "scatter_sg_table_retry",		KSTAT_DATA_UINT64 },
+};
+
+#define	ABDSTAT(stat)		(abd_stats.stat.value.ui64)
+#define	ABDSTAT_INCR(stat, val) \
+	atomic_add_64(&abd_stats.stat.value.ui64, (val))
+#define	ABDSTAT_BUMP(stat)	ABDSTAT_INCR(stat, 1)
+#define	ABDSTAT_BUMPDOWN(stat)	ABDSTAT_INCR(stat, -1)
+
+#define	ABD_SCATTER(abd)	(abd->abd_u.abd_scatter)
+#define	ABD_BUF(abd)		(abd->abd_u.abd_linear.abd_buf)
+#define	abd_for_each_sg(abd, sg, n, i)	\
+	for_each_sg(ABD_SCATTER(abd).abd_sgl, sg, n, i)
+
+/* see block comment above for description */
+int zfs_abd_scatter_enabled = B_TRUE;
+int zfs_abd_scatter_max_order = MAX_ORDER - 1;
+
+static inline size_t
+abd_chunkcnt_for_bytes(size_t size)
+{
+	return (P2ROUNDUP(size, PAGESIZE) / PAGESIZE);
+}
+
+#ifdef _KERNEL
+
+static kstat_t *abd_ksp;
+
+#if !defined(CONFIG_HIGHMEM) && defined(HAVE_SG_FROM_PAGES)
+
+#ifndef __GFP_RECLAIM
+#define	__GFP_RECLAIM		__GFP_WAIT
+#endif
+
+static unsigned long
+abd_alloc_chunk(int nid, gfp_t gfp, unsigned int order)
+{
+	struct page *page;
+
+	page = alloc_pages_node(nid, gfp, order);
+	if (!page)
+		return (0);
+
+	return ((unsigned long) page_address(page));
+}
+
+/*
+ * The goal is to minimize fragmentation by preferentially populating ABDs
+ * with higher order compound pages from a single zone.  Allocation size is
+ * progressively decreased until it can be satisfied without performing
+ * reclaim or compaction.  When nessisary this function will degenerate to
+ * allocating indivial pages and allowing reclaim to satisfy allocations.
+ */
+static void
+abd_alloc_pages(abd_t *abd, size_t size)
+{
+	struct sg_table table;
+	struct page **pages, *page;
+	unsigned long paddr;
+	ssize_t remaining = P2ROUNDUP(size, PAGESIZE);
+	int nr_pages = abd_chunkcnt_for_bytes(size);
+	gfp_t gfp = __GFP_NOWARN | GFP_NOIO;
+	gfp_t gfp_comp = (gfp | __GFP_NORETRY | __GFP_COMP) & ~__GFP_RECLAIM;
+	int order, max_order = zfs_abd_scatter_max_order;
+	int nid = NUMA_NO_NODE;
+	int i = 0, chunks = 0, split = 0, len, error;
+
+	while ((pages = kmalloc(sizeof (*pages) * nr_pages, gfp)) == NULL) {
+		ABDSTAT_BUMP(abdstat_scatter_page_vector_retry);
+		schedule_timeout_interruptible(1);
+	}
+
+	while (remaining > 0) {
+		order = MIN(highbit64(remaining >> PAGE_SHIFT) - 1, max_order);
+		len = 1U << order;
+
+		paddr = abd_alloc_chunk(nid, order ? gfp_comp : gfp, order);
+		if (paddr == 0) {
+			if (order == 0) {
+				ABDSTAT_BUMP(abdstat_scatter_page_alloc_retry);
+				schedule_timeout_interruptible(1);
+			} else {
+				max_order = MAX(0, order - 1);
```

Comment:
I would suggest to revise it to:

```
 max_order = 0;
```
If abd_alloc_chunk() failed to allocate physically contiguous pages, it implies the memory is already fragmented. In that case, instead of trying to squeeze it reducing the number of orders, we should fall back to page-by-page allocation completely.